lib/ddtrace/context.rb

```diff
@@ -194,45 +158,37 @@ def attach_sampling_priority
 
     # Return the start time of the root span, or nil if there are no spans or this is undefined.
     def start_time
-      @mutex.synchronize do
-        return nil if @trace.empty?
-        @trace[0].start_time
-      end
+      return nil if @trace.empty?
+      @trace[0].start_time
     end
 
     # Return the length of the current trace held by this context.
     def length
-      @mutex.synchronize do
-        @trace.length
-      end
+      @trace.length
     end
 
     # Iterate on each span within the trace. This is thread safe.
     def each_span
-      @mutex.synchronize do
-        @trace.each do |span|
-          yield span
-        end
+      @trace.each do |span|
+        yield span
       end
     end
 
     # Delete any span matching the condition. This is thread safe.
     def delete_span_if
-      @mutex.synchronize do
```

Comment:
In removing these mutexes from Context, such as this line, the main concern I'd have is "do contexts get flushed in a multi-threaded way?" Or does flushing a context while having simultaneous writes to it from other processes cause side effects?

I'm a bit concerned with the use of counts in this code, and how it might be more susceptible to issues of state when thread safety no longer exists. e.g.as seen in `check_finished_spans`:

```ruby
# Returns if the trace for the current Context is finished or not. A \Context
# is considered finished if all spans in this context are finished.
def finished?
  @mutex.synchronize do
    return check_finished_spans
  end
end

# Returns if the trace for the current Context is finished or not.
# Low-level internal function, not thread-safe.
 def check_finished_spans
  @finished_spans > 0 && @trace.length == @finished_spans
end
```